File: DP/354_俄罗斯套娃信封问题.py
# 2022-06-24
import bisect
from typing import List


# 会超时
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        n = len(envelopes)
        envelopes.sort(key=lambda x: (x[0], -x[1]))
        dp = [1] * n
        for i in range(1, n):
            for j in range(i):
                if envelopes[i][1] > envelopes[j][1]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)
        


class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        n = len(envelopes)
        
        envelopes.sort(key=lambda x : (x[0], -x[1]))

        dp = [envelopes[0][1]]
        for i in range(1, n):
            if num := envelopes[i][1] > dp[-1]:
                dp.append(num)
            else:
                index = bisect.bisect_left(dp, num)
                dp[index] = num
        
        return len(dp)

File: DP/354_俄罗斯套娃信封问题.cpp
// 2022-06-22


// leetcode 中会超时
class Solution {
public:
    int maxEnvelopes(vector<vector<int>>& envelopes) {
        int n = envelopes.size();
        sort(envelopes.begin(), envelopes.end(), [](const auto &e1, const auto &e2){
            return e1[0] < e2[0] || (e1[0] == e2[0] && e1[1] > e2[1]);
        });
        // dp[i] 表示宽度为 envelopes[i][0], 高度为 envelopes[i][1] 的信封能装的最多信封数 
        vector<int> dp(n, 1);
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (envelopes[i][1] > envelopes[j][1]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }
        return *max_element(dp.begin(), dp.end());
    }
};



// 基于二分查找的动态规划
class Solution {
public:
    int maxEnvelopes(vector<vector<int>>& envelopes) {
        int n = envelopes.size();
        sort(envelopes.begin(), envelopes.end(), [](const auto &e1, const auto &e2){
            return e1[0] < e2[0] || (e1[0] == e2[0] && e1[1] > e2[1]);
        });
        // dp[i] 表示高度为 dp[i] 的信封最多能装 i 封
        vector<int> dp = {envelopes[0][1]};
        for (int i = 1; i < n; i++) {
            int num = envelopes[i][1];
            if (num > dp.back()) {
                dp.push_back(num);
            }
            else {
                auto it = lower_bound(dp.begin(), dp.end(), num);
                *it = num;
            }
        }
        return dp.size();
    }
};